File: QytetetJava/modeloqytetet/Qytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

import QytetetGUI.Dado;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ThreadLocalRandom;

/**
 *
 * @author csp98
 */
public class Qytetet {

    private Sorpresa cartaActual;
    private ArrayList<Jugador> jugadores;
    private ArrayList<Sorpresa> mazo;
    private Jugador jugadorActual;
    private Tablero tablero;
    private Dado dado;

    public static final int MAX_JUGADORES = 4;
    protected static final int MAX_CARTAS = 10;
    protected static final int MAX_CASILLAS = 20;
    public static final int PRECIO_LIBERTAD = 200;
    public static final int SALDO_SALIDA = 1000;

    private static Qytetet INSTANCE;

    public static Qytetet getInstance(ArrayList<String> jugadores) {
        if (INSTANCE == null) {
            INSTANCE = new Qytetet(jugadores);
        }
        return INSTANCE;
    }

    private Qytetet(ArrayList<String> nombreJugadores) {
        inicializarJuego(nombreJugadores);
    }

    private void inicializarCartasSorpresa() {
        mazo = new ArrayList();
        mazo.add(new Sorpresa("Te conviertes en especulador.\nTu fianza es de 5000 €", 5000, TipoSorpresa.CONVERTIRSE));
        mazo.add(new Sorpresa("Te conviertes en especulador.\nTu fianza es de 3000 €", 3000, TipoSorpresa.CONVERTIRSE));
        mazo.add(new Sorpresa("Se te vienen encima los exámenes finales e intentas salvar\n"
                + "el curso con una academia. Paga 100€ para clases", -100, TipoSorpresa.PAGARCOBRAR));
        mazo.add(new Sorpresa("Sacas matrícula de honor en una asignatura.\nRecibes 100€", 100, TipoSorpresa.PAGARCOBRAR));
        mazo.add(new Sorpresa("Suspendes una asignatura y no te guardan las prácticas.\nVuelve al punto de salida", 0, TipoSorpresa.IRACASILLA));
        mazo.add(new Sorpresa("Pierdes las llaves de casa y pasas una noche en la casilla 8.", 8, TipoSorpresa.IRACASILLA));
        mazo.add(new Sorpresa("Te pillan copiando una práctica. Vas a la cárcel", 10, TipoSorpresa.IRACASILLA));
        mazo.add(new Sorpresa("Los inquilinos de tus propiedades pagan su estancia.\nRecibe 100€ por casa y 200 por hotel", +100, TipoSorpresa.PORCASAHOTEL));
        mazo.add(new Sorpresa("Das una macrofiesta en tus propiedades pero la cosa se desmadra.\n"
                + "Paga 75€ por cada casa que tengas y 150€ por hotel.", -75, TipoSorpresa.PORCASAHOTEL));
        mazo.add(new Sorpresa("Cada jugador te hace un regalo de 75€ por tu cumpleaños.", +75, TipoSorpresa.PORJUGADOR));
        mazo.add(new Sorpresa("No te cuentan las prácticas en la convocatoria extraordinaria y tu nota\n"
                + "se queda en 4,89. Lo siento, pero pagas segunda matrícula (160 )", -160, TipoSorpresa.PORJUGADOR));
        mazo.add(new Sorpresa("Tu mejor amigo paga la fianza.\n"
                + "Eres libre de nuevo.", 0, TipoSorpresa.SALIRCARCEL));
        barajarCartas();
    }

    private void barajarCartas() {
        int veces = (int) (Math.random() * 10 + 1); //Entre 1 y 10
        for (int i = 0; i < veces; i++) {
            int j = (int) (Math.random() * mazo.size() / 2);  //Entre 0 y la mitad
            int k = (int) (Math.random() * mazo.size() + mazo.size() / 2); //Entre la mitad y el final
            Sorpresa temp = new Sorpresa(mazo.get(i));
            mazo.set(i, mazo.get(j));
            mazo.set(j, temp);
        }
    }

    private void inicializarJugadores(ArrayList<String> nombres) {
        jugadores = new ArrayList();
        for (int i = 0; i < nombres.size(); i++) {
            jugadores.add(new Jugador(nombres.get(i)));
            jugadores.get(i).setCasillaActual(tablero.obtenerCasillaNumero(0)); //Establecemos la casilla inicial de salida.
        }
    }

    private void inicializarTablero() {
        tablero = new Tablero();
    }

    public Sorpresa getCartaActual() {
        return cartaActual;
    }

    public Jugador getJugadorActual() {
        return jugadorActual;
    }

    public void siguienteJugador() {
        int posicionActual, posicionNueva;
        posicionActual = jugadores.lastIndexOf(jugadorActual);
        posicionNueva = (posicionActual + 1) % jugadores.size();
        jugadorActual = jugadores.get(posicionNueva);
    }

    public boolean aplicarSorpresa() {
        boolean tienePropietario = false;
        Casilla nuevaCasilla;
        boolean esCarcel;
        if (cartaActual.getTipo() == TipoSorpresa.PAGARCOBRAR) {
            jugadorActual.modificarSaldo(cartaActual.getValor());
        } else if (cartaActual.getTipo() == TipoSorpresa.IRACASILLA) {
            esCarcel = tablero.esCasillaCarcel(cartaActual.getValor());
            if (esCarcel) {
                encarcelarJugador();
            } else {
                nuevaCasilla = tablero.obtenerCasillaNumero(cartaActual.getValor());
                tienePropietario = jugadorActual.actualizarPosicion(nuevaCasilla);
            }
        } else if (cartaActual.getTipo() == TipoSorpresa.PORCASAHOTEL) {
            jugadorActual.pagarCobrarPorCasaYHotel(cartaActual.getValor());
        } else if (cartaActual.getTipo() == TipoSorpresa.PORJUGADOR) {
            for (int i = 0; i < jugadores.size(); i++) {
                if (jugadores.get(i) != jugadorActual) {
                    jugadores.get(i).modificarSaldo(cartaActual.getValor());
                    jugadorActual.modificarSaldo(-cartaActual.getValor());
                }
            }
        } else if (cartaActual.getTipo() == TipoSorpresa.CONVERTIRSE) {
            int fianza = cartaActual.getValor();
            Especulador nuevo = jugadorActual.convertirme(fianza);
            int indice = jugadores.indexOf(jugadorActual);
            jugadores.remove(jugadorActual);
            jugadores.add(indice, nuevo);
            jugadorActual = nuevo;
            for (TituloPropiedad p : jugadorActual.propiedades) {
                p.setPropietario(jugadorActual);
            }
        }
        mazo.remove(cartaActual);
        if (cartaActual.getTipo() == TipoSorpresa.SALIRCARCEL) {
            jugadorActual.setCartaLibertad(cartaActual);
        } else {
            mazo.add(cartaActual);
        }
        return tienePropietario;
    }

    public boolean cancelarHipoteca(Casilla casilla) {
        boolean puedoDeshipotecar = false;
        boolean estaHipotecada;
        int aPagar;
        estaHipotecada = ((Calle) casilla).estaHipotecada();
        if (estaHipotecada) {
            puedoDeshipotecar = jugadorActual.puedoPagarHipoteca(casilla);
            if (puedoDeshipotecar) {
                aPagar = ((Calle) casilla).cancelarHipoteca();
                jugadorActual.modificarSaldo(-aPagar);
            }
        }
        return puedoDeshipotecar;
    }

    public boolean comprarTituloPropiedad() {
        return jugadorActual.comprarTitulo();
    }

    public boolean edificarCasa(Casilla casilla) {
        boolean puedoEdificar = false;
        if (casilla.soyEdificable()) {
            boolean sePuedeEdificar;
            sePuedeEdificar = ((Calle) casilla).sePuedeEdificarCasa(jugadorActual.getFactorEspeculador());
            if (sePuedeEdificar) {
                puedoEdificar = jugadorActual.puedoEdificarCasa(casilla);
                if (puedoEdificar) {
                    int costeEdificarCasa;
                    costeEdificarCasa = ((Calle) casilla).edificarCasa();
                    jugadorActual.modificarSaldo(-costeEdificarCasa);
                }
            }
        }
        return puedoEdificar;
    }

    public boolean edificarHotel(Casilla casilla) {
        boolean puedoEdificar = false;
        if (casilla.soyEdificable()) {
            boolean sePuedeEdificar;
            sePuedeEdificar = ((Calle) casilla).sePuedeEdificarHotel(jugadorActual.getFactorEspeculador());
            if (sePuedeEdificar) {
                puedoEdificar = jugadorActual.puedoEdificarHotel(casilla);
            }
            if (puedoEdificar) {
                int costeEdificarHotel;
                costeEdificarHotel = ((Calle) casilla).edificarHotel();
                jugadorActual.modificarSaldo(-costeEdificarHotel);
            }
        }
        return puedoEdificar;
    }

    public boolean hipotecarPropiedad(Casilla casilla) {
        boolean puedoHipotecarPropiedad = false;
        if (casilla.soyEdificable()) {
            boolean sePuedeHipotecar;
            sePuedeHipotecar = !((Calle) casilla).estaHipotecada();
            if (sePuedeHipotecar) {
                puedoHipotecarPropiedad = jugadorActual.puedoHipotecar(casilla);
                if (puedoHipotecarPropiedad) {
                    int cantidadRecibida;
                    cantidadRecibida = ((Calle) casilla).hipotecar();
                    jugadorActual.modificarSaldo(cantidadRecibida);
                }
            }
        }
        return puedoHipotecarPropiedad;
    }

    public void inicializarJuego(ArrayList<String> jugadores) {
        inicializarTablero();
        inicializarCartasSorpresa();
        inicializarJugadores(jugadores);
        salidaJugadores();
    }

    public boolean intentarSalirCarcel(MetodoSalirCarcel metodo) {
        boolean libre = false;
        boolean tengoSaldo;
        int valorDado;
        dado = Dado.getInstance();
        if (metodo == MetodoSalirCarcel.TIRANDODADO) {
            valorDado = dado.nextNumber();
            libre = valorDado > 5;
        } else if (metodo == MetodoSalirCarcel.PAGANDOLIBERTAD) {
            tengoSaldo = jugadorActual.pagarLibertad(PRECIO_LIBERTAD);
            libre = tengoSaldo;
            if (tengoSaldo) {
                jugadorActual.modificarSaldo(-PRECIO_LIBERTAD);
            }
        }
        if (libre) {
            jugadorActual.setEncarcelado(false);
        }
        return libre;
    }

    public boolean jugar() {
        int valorDado;
        Casilla casillaPosicion;
        Casilla nuevaCasilla;
        boolean tienePropietario;
        dado = Dado.getInstance();
        valorDado = dado.nextNumber();
        casillaPosicion = jugadorActual.getCasillaActual();
        nuevaCasilla = tablero.obtenerNuevaCasilla(casillaPosicion, valorDado);
        tienePropietario = jugadorActual.actualizarPosicion(nuevaCasilla);
        if (!nuevaCasilla.soyEdificable()) {
            if (nuevaCasilla.getTipoCasilla() == TipoCasilla.JUEZ) {
                encarcelarJugador();
            } else if (nuevaCasilla.getTipoCasilla() == TipoCasilla.SORPRESA) {
                cartaActual = mazo.get(0);
            }
        }
        return tienePropietario;
    }

    public Map<String, Integer> obtenerRanking() {
        Map<String, Integer> ranking = new HashMap();
        int capital;
        String nombre;
        for (int i = 0; i < jugadores.size(); i++) {
            capital = jugadores.get(i).obtenerCapital();
            nombre = jugadores.get(i).getNombre();
            ranking.put(nombre, capital);
        }
        return ranking;
    }

    public ArrayList<Casilla> propiedadesHipotecadasJugador(boolean hipotecadas) {
        ArrayList<TituloPropiedad> titulos = jugadorActual.obtenerPropiedadesHipotecadas(hipotecadas);
        ArrayList<Casilla> aDevolver = new ArrayList();
        for (int i = 0; i < titulos.size(); i++) {
            aDevolver.add(titulos.get(i).getCasilla());
        }
        return aDevolver;
    }

    public boolean venderPropiedad(Casilla casilla) {
        boolean puedoVender = false;
        if (casilla.soyEdificable()) {
            puedoVender = jugadorActual.puedoVenderPropiedad(casilla);
            if (puedoVender) {
                jugadorActual.venderPropiedad(casilla);
            }
        }
        return puedoVender;
    }

    private void encarcelarJugador() {
        Casilla casillaCarcel;
        Sorpresa carta;
        if (!jugadorActual.tengoCartaLibertad()) {
            casillaCarcel = tablero.getCarcel();
            jugadorActual.irACarcel(casillaCarcel);
        } else {
            carta = jugadorActual.devolverCartaLibertad();
            mazo.add(carta);
        }
    }

    private void salidaJugadores() {
        for (int i = 0; i < jugadores.size(); i++) {
            jugadores.get(i).setCasillaActual(tablero.obtenerCasillaNumero(0));
        }
        int player = ThreadLocalRandom.current().nextInt(0, jugadores.size() - 1);
        jugadorActual = jugadores.get(player);
    }

    public ArrayList<Jugador> getJugadores() {
        return jugadores;
    }

    public Tablero getTablero() {
        return tablero;
    }

    public ArrayList<Sorpresa> getSorpresas() {
        return mazo;
    }
    public int getSaldoSalida(){
        return SALDO_SALIDA;
    }
}

File: QytetetRuby/Qytetet.rb
#encoding: utf-8

# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require "singleton"



require_relative 'Casilla'
require_relative 'TipoCasilla'
require_relative 'MetodoSalirCarcel'
require_relative 'Tablero'
require_relative 'TipoSorpresa'
require_relative 'Sorpresa'
require_relative 'Jugador'
require_relative 'Especulador'
require_relative 'Dado'

module ModeloQytetet
  class Qytetet
    include Singleton
    def initialize
      @CartaActual=nil
      @jugadores=Array.new
      @mazo=Array.new
      @jugadorActual=nil
      @tablero=nil
      @dado=Dado.new
    end

    public

    attr_reader :MAX_JUGADORES,:MAX_CARTAS,:MAX_CASILLAS,:PRECIO_LIBERTAD,:SALDO_SALIDA
    @@MAX_JUGADORES=4
    @@MAX_CARTAS=10
    @@MAX_CASILLAS=20
    @@PRECIO_LIBERTAD=200
    @@SALDO_SALIDA=1000
    attr_reader :cartaActual , :jugadorActual , :jugadores, :tablero , :mazo
    
    def inicializarJugadores (nombres)
      i=0
      while i<nombres.size
        @jugadores<<Jugador.new(nombres[i])
        i=i+1
      end
    end
    
    def barajarCartas
      veces= rand(10)+1
      i=0
      while i!=veces
        j=rand(@mazo.size)
        k=rand(@mazo.size)
        temp=Sorpresa.new(@mazo[i].descripcion,@mazo[i].valor,@mazo[i].tipo)
        @mazo[i]=@mazo[j]
        @mazo[j]=temp
        i+=1
      end
    end
    
    def aplicarSorpresa
      tienePropietario=false
      if @cartaActual.tipo == TipoSorpresa::PAGARCOBRAR
        @jugadorActual.modificarSaldo(@cartaActual.valor)
      elsif @cartaActual.tipo == TipoSorpresa::IRACASILLA
        esCarcel=@tablero.esCasillaCarcel(@cartaActual.valor)
        if esCarcel
          encarcelarJugador
        else
          nuevaCasilla=@tablero.obtenerCasillaNumero(@cartaActual.valor)
          tienePropietario=@jugadorActual.actualizarPosicion(nuevaCasilla)
        end
      elsif @cartaActual.tipo == TipoSorpresa::PORCASAHOTEL
        @jugadorActual.pagarCobrarPorCasaYHotel(@cartaActual.valor)
      elsif @cartaActual.tipo == TipoSorpresa::PORJUGADOR
        i=0
        while i < @jugadores.size
          if @jugadores[i] != @jugadorActual
            @jugadores[i].modificarSaldo(@cartaActual.valor)
            @jugadorActual.modificarSaldo(-@cartaActual.valor)
          end
        end
      elsif @cartaActual.tipo == TipoSorpresa::CONVERTIRSE
        fianza=@cartaActual.valor
        nuevo=@jugadorActual.convertirme(fianza)
        @jugadores[@jugadores.index(@jugadorActual)]=nuevo
        @jugadorActual=nuevo
        for p in @jugadorActual.propiedades
          p.propietario=@jugadorActual
        end
      end
      @mazo.delete(@cartaActual)
      if@cartaActual.tipo == TipoSorpresa::SALIRCARCEL
        @jugadorActual.cartaLibertad=@cartaActual
      else
        @mazo<<@cartaActual
      end
      return tienePropietario
    end

    def cancelarHipoteca(casilla)
      puedoDeshipotecar=false
      estaHipotecada=casilla.estaHipotecada
      if estaHipotecada
        puedoDeshipotecar=@jugadorActual.puedoPagarHipoteca(casilla)
        if puedoDeshipotecar
          aPagar = casilla.cancelarHipoteca
          @jugadorActual.modificarSaldo(-aPagar)
        end
      end
      return puedoDeshipotecar
    end

    def comprarTituloPropiedad
      puedoComprar=@jugadorActual.comprarTitulo
      return puedoComprar
    end

    def edificarCasa(casilla)
      puedo_edificar = false
      if casilla.soyEdificable
        if casilla.sePuedeEdificarCasa(@jugadorActual.factorEspeculador)
          puedo_edificar = @jugadorActual.puedoEdificarCasa(casilla)
          if (puedo_edificar)
            costeEdificarCasa = casilla.edificarCasa
            @jugadorActual.modificarSaldo(-costeEdificarCasa)
          end
        end
      end
      return puedo_edificar
    end
    
    def edificarHotel(casilla)
      puedo_edificar = false
      if casilla.soyEdificable
        if casilla.sePuedeEdificarHotel(@jugadorActual.factorEspeculador)
          puedo_edificar = @jugadorActual.puedoEdificarHotel(casilla)
          if puedo_edificar
            costeEdificarHotel = casilla.edificarHotel
            @jugadorActual.modificarSaldo(-costeEdificarHotel)
          end
        end
      end
      return puedo_edificar
    end

    def hipotecarPropiedad(casilla)
      puedoHipotecarPropiedad=false
      if casilla.soyEdificable
        sePuedeHipotecar= !casilla.estaHipotecada
        if sePuedeHipotecar
          puedoHipotecar= @jugadorActual.puedoHipotecar(casilla)
          if puedoHipotecar
            cantidadRecibida=casilla.hipotecar
            @jugadorActual.modificarSaldo(cantidadRecibida)
            puedoHipotecarPropiedad=puedoHipotecar
          end
        end
      end
      return puedoHipotecarPropiedad
    end

    def inicializarJuego(nombres)
      inicializarCartasSorpresa
      inicializarTablero
      inicializarJugadores(nombres)
      salidaJugadores
    end

    def intentarSalirCarcel (metodo)
      libre=false
      if metodo == MetodoSalirCarcel::TIRANDODADO
        valorDado=@dado.tirar
      elsif metodo == MetodoSalirCarcel::PAGANDOLIBERTAD
        tengoSaldo=@jugadorActual.pagarLibertad(-@@PRECIO_LIBERTAD)
        libre=tengoSaldo
        if libre
          @jugadorActual.encarcelado=false
        end
      end
      return libre
    end

    def jugar
      valorDado=@dado.tirar
      casillaPosicion=@jugadorActual.casillaActual
      nuevaCasilla=@tablero.obtenerNuevaCasilla(casillaPosicion,valorDado)
      tienePropietario=@jugadorActual.actualizarPosicion(nuevaCasilla)
      if !nuevaCasilla.soyEdificable
        if nuevaCasilla.tipo == TipoCasilla::JUEZ
          encarcelarJugador
        elsif nuevaCasilla.tipo == TipoCasilla::SORPRESA
          @cartaActual=@mazo[0]
        end
      end
      return tienePropietario
    end
    def obtenerRanking
      ranking = Hash.new
      for j in @jugadores
        ranking[j.nombre] = j.obtenerCapital
      end
      return ranking
    end

    def propiedadesHipotecadasJugador(hipotecadas)
      aux = Array.new
      for i in @jugadorActual.propiedades.size
        if (@jugadorActual.propiedades[i].hipotecada == hipotecadas)
          aux << @jugadorActual.propiedades[i]
        end
      end

      return aux
    end

    def  siguienteJugador
      nuevo =( @jugadores.index(@jugadorActual) + 1) % @jugadores.size
      @jugadorActual=@jugadores[nuevo]
    end


    def venderPropiedad(casilla)
      puedoVender = false
      if casilla.soyEdificable
        puedoVender = @jugadorActual.puedoVenderPropiedad(casilla);
        if puedoVender
          @jugadorActual.venderPropiedad(casilla)
        end
      end
      return puedoVender
    end


    private

    def encarcelarJugador
      if !@jugadorActual.tengoCartaLibertad
        casillaCarcel=@tablero.carcel
        @jugadorActual.irACarcel(casillaCarcel)
      else
        carta=@jugadorActual.devolverCartaLibertad
        @mazo<<carta
      end
    end

    def inicializarTablero
      @tablero=Tablero.new
    end

    def inicializarCartasSorpresa
      @mazo<<Sorpresa.new("Te conviertes en especulador. Tu fianza es de 5000€",5000,TipoSorpresa::CONVERTIRSE)
      @mazo<<Sorpresa.new("Se te vienen encima los examenes finales e intentas salvar el curso con una academia. Paga 100 euros para clases",-100,TipoSorpresa::PAGARCOBRAR)
      @mazo<<Sorpresa.new("Te pillan copiando una practica. Ve a la carcel",9,TipoSorpresa::IRACASILLA)
      @mazo<<Sorpresa.new("Pierdes las llaves de casa y pasas una noche en la casilla 8. Ve a la casilla 8",8, TipoSorpresa::IRACASILLA);
      @mazo<<Sorpresa.new("Suspendes una asignatura y no te guardan las prácticas. Vuelve al punto de salida",0, TipoSorpresa::IRACASILLA)
      @mazo<<Sorpresa.new("Das una macrofiesta en tus propiedades pero la cosa se desmadra. Paga 75 euros por casa y 150 por hotel.",-75,TipoSorpresa::PORCASAHOTEL)
      @mazo<<Sorpresa.new("No te cuentan las practicas en la convocatoria extraordinaria y tu nota se queda en 4.89. Lo siento, pagas segunda matricula",-160,TipoSorpresa::PORJUGADOR)
      @mazo<<Sorpresa.new("Aquel chaval que aprobo programacion gracias a ti paga tu fianza. Eres libre de nuevo",0,TipoSorpresa::SALIRCARCEL)
      @mazo<<Sorpresa.new("Sacas matrícula de honor en una asignatura y la universidad te devuelve el pago de una matrícula. Recibes 100€", 100, TipoSorpresa::PAGARCOBRAR);
      @mazo<<Sorpresa.new("Los inquilinos de tus propiedades pagan su estancia. Recibe 100€ por casa y 200 por hotel", +100, TipoSorpresa::PORCASAHOTEL)
      @mazo<<Sorpresa.new("Cada jugador te hace un regalo de 75€ por ser buena gente. Recibe 75€ por cada jugador", +75, TipoSorpresa::PORJUGADOR)
      @mazo<<Sorpresa.new("Te conviertes en especulador. Tu fianza es de 3000€",3000,TipoSorpresa::CONVERTIRSE)
      barajarCartas
    end

    def salidaJugadores

      aleatorio = rand(@jugadores.size)

      for j in @jugadores
        j.casillaActual = @tablero.obtenerCasillaNumero(0)
        j.modificarSaldo(7500)
      end

      @jugadorActual = @jugadores[aleatorio]
    end

  end
end
    